docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,309 @@
+or the descriptor as it looks after the `CREATE INDEX` statement. If
+the `INSERT` statement does not see the `CREATE INDEX` statement then
+it will merrily insert the values into `test` without creating the
+entry for the `foo` index, thus leaving the table data in an
+inconsistent state.
+
+It is easy to show similar problematic sequences of operations for
+`DELETE`, `UPDATE` and `SELECT`. Fortunately, a solution exists in the
+literature: break up the schema change into discrete steps for which
+any two consecutive versions of the schema are safe for concurrent
+use. For example, to add an index to a table we would perform the
+following steps:
+
+* Add the index to the table descriptor but mark it as delete-only:
+  update and delete operations which would delete entries for the
+  index do so, but insert and update operations which would add new
+  elements and read operations do not use the index.
+* Wait for the table descriptor change to propagate to all nodes in
+  the cluster.
+* Mark the index as write-only: insert, update and delete operations
+  would add entries for the index, but read operations would not use
+  the index.
+* Wait for the table descriptor change to propagate to all nodes in
+  the cluster.
+* Backfill the index.
+* Mark the index in the table descriptor as read-write.
+* Wait for the table descriptor change to propagate to all nodes in
+  the cluster.
+
+This RFC is focused on how to wait for the table descriptor change
+propagate to all nodes in the cluster. Additionally, we need
+mechanisms to ensure that multiple updates to a table descriptor are
+not performed concurrently and that write operations never commit once
+a table descriptor is too old (i.e. not one of the two most recent
+versions).
+
+At a high-level, we want to provide read and write leases for table
+descriptors. When a schema change is to be performed, the operation
+first needs to grab a write lease for the descriptor. The write lease
+ensures that no concurrent schema change will be performed and that
+there is only one active version of the descriptor in the cluster.
+Note that these are essentially the same constraint: there are only
+two active versions of a descriptor during a schema change.
+
+Before using a table descriptor for a DML operation (i.e. `SELECT`,
+`INSERT`, `UPDATE`, `DELETE`, etc), the operation needs to obtain a
+read lease for the descriptor. The lease will be guaranteed valid for
+a significant period of time (on the order of minutes) and the lease
+period can be extended as long as a scheme change is not
+performed. When the operation completes it will release the read
+lease.
+
+# Detailed design
+
+Leases will be managed and dispensed by a centralized lease
+service. Initially there will be a single lease service master, but a
+future enhancement to shard by descriptor ID should be
+straightforward.
+
+The lease master will be "elected" by performing a conditional put to
```

Comment:
If we only have a single shard then yeah, using the leader lease on the `/lease-master` key would be simpler. If we have multiple shards (or want to prepare for that eventuality), how would you ensure that the "leader keys" are distributed across multiple ranges?